### PR TITLE
Feature - CircularFlow by Rodrigo Dominguez

### PR DESCRIPTION
**Pull request by rodrigomartind**
# CircularFlow
I add a class called CircularFlow which allows us to group views with circular references.

`constraint_referenced_ids` : It receives id's of the views that will add the references (Eg. `"view2, view3, view4,view5,view6"`).

`circularflow_viewCenter` : It receives the id of the view of the center where the views received in _constraint_referenced_ids_ (Eg. `"view1"`)

`circularflow_angles` : Receive the angles that you will assign to each view (Eg. `"45,90,135,180,225"`)

`circularflow_radiusInDP` : Receive the radios in DP that you will assign to each view (Eg. `"90,100,110,120,130"`)

## How add a view at runtime? 
With method `addViewToCircularFlow` you can add new views to CircularFlow.
```
 /**
 * Add a view to the CircularFlow. The referenced view need to be a child of the container parent.
 * The view also need to have its id set in order to be added.
 * The views previous need to have its radius and angle set in order to be added correctly a new view.
 * @param view
 * @param radius
 * @param angle
 * @return
 */
```

Eg.
```
findViewById<View>(R.id.view7).setOnClickListener {
            findViewById<CircularFlow>(R.id.circularFlow).addViewToCircularFlow(
                it, 140, 200F
            )
        }
```
## Demo add view in emulator

https://user-images.githubusercontent.com/24611045/107442324-ad3eee80-6b15-11eb-89b5-7f3221f2295d.mp4

## How remove a view at runtime? 
With method `removeViewFromCircularFlow` you can remove a view from CircularFlow.
```
 /**
     * Remove a given view from the CircularFlow.
     *
     * @param view
     */
```

Eg.
```
findViewById<View>(R.id.view2).setOnClickListener {
            findViewById<CircularFlow>(R.id.circularFlow).removeViewFromCircularFlow(
                it
            )
        }
```
## Demo remove view in emulator

https://user-images.githubusercontent.com/24611045/107656012-80e4b880-6c62-11eb-924f-b0fa1244a2bf.mp4

## Demos in AS:
https://user-images.githubusercontent.com/24611045/106652697-9aa44280-6574-11eb-819b-cb09f012f748.mov

https://user-images.githubusercontent.com/24611045/106653120-25853d00-6575-11eb-86ab-dca1a1c85ed9.mov

### CODE REVIEW

**Comment by rodrigomartind** [comment]
@googlebot I consent.

**Comment by rodrigomartind** [comment]
@googlebot I consent

**Comment by rodrigomartindMeli** [comment]
@googlebot I consent.

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+package androidx.constraintlayout.helper.widget;
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:1]: missing (c) notice

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        if (attrs != null) {
+            TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.ConstraintLayout_Layout);
+            final int N = a.getIndexCount();
+
+            for (int i = 0; i < N; i++) {
+                int attr = a.getIndex(i);
+                if (attr == R.styleable.ConstraintLayout_Layout_circularflow_viewCenter) {
+                    mViewCenter = a.getResourceId(attr, 0);
+                }
+            }
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:45]: add a.recycle();

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintHelper.java [hunk 3]
```diff
@@ -46,10 +46,27 @@
      * @hide
      */
     protected int[] mIds = new int[32];
+    /**
+     * @hide
+     */
+    protected float[] mAngles = new float[32];
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintHelper.java:52]: those added variables should be in CircularFlow, not in ConstraintHelper (mAngles, mRadius, etc.)

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintHelper.java [hunk 4]
```diff
@@ -112,6 +138,12 @@ protected void init(AttributeSet attrs) {
                 } else if (attr == R.styleable.ConstraintLayout_Layout_constraint_referenced_tags) {
                     mReferenceTags = a.getString(attr);
                     setReferenceTags(mReferenceTags);
+                } else if (attr == R.styleable.ConstraintLayout_Layout_circularflow_angles) {
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintHelper.java:141]: Similar -- you should add that in CircularFlow

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    @Override
+    public void onAttachedToWindow() {
+        super.onAttachedToWindow();
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:51]: This set constraints only on attached to window -- what about if you want to add views to CircularFlow at runtime?

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java (hunk 1 above)
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:16]: Done ✅

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintHelper.java (hunk 3 above)
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintHelper.java]: Done ✅

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintHelper.java (hunk 4 above)
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/widget/ConstraintHelper.java]: Done ✅

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java (hunk 5 above)
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:112]: I added a new method `addViewToCircularFlow` to add views to CircularFlow.

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java (hunk 2 above)
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:105]: Done ✅

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    /**
+     * Add a view to the CircularFlow. The referenced view need to be a child of the container parent.
+     * The view also need to have its id set in order to be added.
+     * The views previous need to have its radius and angle set in order to be added correctly a new view.
+     * @param view
+     * @param radius
+     * @param angle
+     * @return
+     */
+    public void addViewToCircularFlow(View view, int radius, float angle) {
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:161]: should you have the corresponding removeView() ? or is the one in ConstraintHelper doing enough?

**Review by camaelon — CHANGES_REQUESTED** [review]
looks good! if you can address the minor comments I added we can submit this. Thanks!

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
```
[the remark this answers is not in the document]
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java]: Done ✅

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+import android.util.AttributeSet;
+import android.util.Log;
+import android.view.View;
+import androidx.constraintlayout.widget.ConstraintLayout;
+import androidx.constraintlayout.widget.ConstraintSet;
+import androidx.constraintlayout.widget.R;
+import androidx.constraintlayout.widget.VirtualLayout;
+import java.util.Arrays;
+
+public class CircularFlow extends VirtualLayout {
```
[the remark this answers is not in the document]
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:62]: Done ✅

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java (hunk 6 above)
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:194]: I added a `removeViewFromCircularFlow` where remove view with `removeView` from ConstraintHelper, and remove radius/angles for this view.

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    /**
+     * @hide
+     */
+    private float[] removeAngle(float[] angles, int index) {
+        if (angles == null
+                || index < 0
+                || index >= angles.length) {
+            return angles;
+        }
+        float[] anotherArray = new float[32];
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:239]: er -- you probably want to size this array according to the current size of the angles array, which technically could be more than 32 no?

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+    /**
+     * @hide
+     */
+    private int[] removeRadius(int[] radius, int index) {
+        if (radius == null
+                || index < 0
+                || index >= radius.length) {
+            return radius;
+        }
+        int[] anotherArray = new int[32];
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:260]: same comment as above

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java (hunk 9 above)
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java]: Done ✅

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java (hunk 10 above)
Comment (rodrigomartind) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java]: Done ✅

**Comment by rodrigomartind** [comment]
Thanks for your review Nicolas 🚀

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+        mRadius[mCountRadius - 1] = (int) (radius * myContext.getResources().getDisplayMetrics().density);
+        anchorReferences();
+    }
+
+    /**
+     * Remove a given view from the CircularFlow.
+     *
+     * @param view
+     */
+    public void removeViewFromCircularFlow(View view){
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:213]: minor nit, but if there's no additional arguments, why not override removeView() ?

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+        }
+        anchorReferences();
+    }
+
+
+    /**
+     * @hide
+     */
+    private float[] removeAngle(float[] angles, int index) {
+        if (angles == null
```
Comment (camaelon) [inline constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:239]: small suggestion: refactor in a utility function to remove an element of a given array and call it from removeAngle() and removeRadius()

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java (hunk 11 above)
Comment (rodrigomartindMeli) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java]: Refactor done ✅

File: constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java (hunk 12 above)
Comment (rodrigomartindMeli) [reply constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java:237]: Refactor done ✅

**Review by camaelon — APPROVED** [review]

**Pull request merged**